util/geoutils.py

```diff
@@ -0,0 +1,354 @@
+
+    # Load the shapefile
+    shapefile = gpd.read_file(shapefile_path)
+
+    with rasterio.open(raster_path) as src:
+        # Ensure the shapefile is in the same CRS as the raster
+        shapefile = shapefile.to_crs(src.crs)
+
+        # Clip the raster with the shapefile
+        geoms = [mapping(shape) for shape in shapefile.geometry]
+        out_image, out_transform = mask(src, geoms, crop=True)
+
+        # Filter out invalid values
+        for invalid_value in invalid_values:
+            out_image = np.where(out_image == invalid_value, np.nan, out_image)
+
+        out_meta = src.meta.copy()
+
+    # Update metadata to reflect the number of layers, new transform, and new dimensions
+    out_meta.update({
+        "driver": "GTiff",
+        "height": out_image.shape[1],
+        "width": out_image.shape[2],
+        "transform": out_transform,
+        # Ensure dtype is float to accommodate NaN values
+        "dtype": 'float32'
+    })
+
+    output_path = os.path.join(
+        output_clip, f"clipped_{file_name}")  # Modify as needed
+    # Write the clipped and filtered raster to file
+    with rasterio.open(output_path, "w", **out_meta) as dest:
+        dest.write(out_image)
+
+
+def clipMultipleRasters(raster_paths, shapefile_path, invalid_values=None):
+
+    for raster_path in tqdm(raster_paths, total=len(raster_paths), desc="Clipping Rasters"):
+        clipRasterWithShapefile(raster_path, shapefile_path, invalid_values)
+
+    print(f"Raster clipped successfully and saved to {output_clip}")
+
+    return loadData(output_clip)
+
+
+def extract_date_from_filename(filename):
+    pattern = r'\d{8}'
+    match = re.search(pattern, filename)
+    date_str = match.group()
+    return date_str
+
+
+def aggregate_rasters_conditional(predictor_paths, shapefile_path, output_csv_path, crop_mask_path=None, use_crop_mask=False, predictor_name="Value"):
+    data_list = []
+    shape_file = gpd.read_file(shapefile_path)
+
+    if use_crop_mask:
+        if crop_mask_path is None:
+            raise ValueError(
+                "Crop mask path must be provided if use_crop_mask is True.")
+        mask_src = rasterio.open(crop_mask_path)
+        # Dynamically read the no data value from the crop mask metadata
+        no_data_value = mask_src.nodata
+
+    for raster in tqdm(predictor_paths, desc="Processing rasters", unit="raster"):
+        file_name = os.path.basename(raster)
+        date_str = extract_date_from_filename(file_name)
+
+        with rasterio.open(raster) as src:
+            for _, row in shape_file.iterrows():
+                geom = row.geometry
+                out_image, _ = mask(src, [geom], crop=True)
+
+                if use_crop_mask:
+                    mask_image, _ = mask(mask_src, [geom], crop=True)
+                    if no_data_value is not None:
+                        # Mask out no data values in the crop mask
+                        valid_mask = (mask_image != no_data_value)
```

Comment:
@AbdelrahmanAmr3 if the nodata value in the crop mask happen to be nan, then (mask_iimage!=no_data_value) will always return true for nans. for such cases, perhaps ~np.isnan(mask_iimage)?